storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3Storage.java

```diff
@@ -47,9 +47,12 @@ public void configure(final Map<String, ?> configs) {
 
     @Override
     public void upload(final InputStream inputStream, final String key) throws StorageBackendException {
-        try (final var out = s3OutputStream(key)) {
+        final S3MultiPartOutputStream out = s3OutputStream(key);
+        try {
             inputStream.transferTo(out);
+            out.close();
```

Comment:
Why not try-finally?

Reply:
Want to avoid throwing IllegalStateException as stream is already closed after calling abort.

Reply:
So abort in the catch block guarantees that it will be closed? That should be documented at least I guess

Reply:
Agree, adding some notes to the class.

Reply:
Is it covered with tests also?

Reply:
Good catch. Adding a test case on S3StorageTest

Reply:
> Want to avoid throwing IllegalStateException as stream is already closed after calling abort.

If throwing `IllegalStateException` breaks basic language idioms, probably it's not a good idea to do this. In most (all?) standard `InputStream`s `close` is idempotent and can be called multiple times. We probably should follow

Reply:
@ivanyu good point. I think close validation can be relaxed and only log and avoid side effects instead of throwing exceptions.

Reply:
Was thinking about exactly what @ivanyu said in the last comment.